### PR TITLE
window: fix win_free_all() heap-use-after-free

### LINKED ISSUE #10838: heap-use-after-free on exit with aucmd_win != NULL
- `nvim --version` ([master](https://github.com/neovim/neovim/tree/34e416667378a610943f614b97da9b865d622806)):
```
NVIM v0.4.0-1801-g34e416667
Build type: Debug
LuaJIT 2.0.5
Compilation: /usr/bin/clang -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=always -fno-sanitize-recover=all -fno-omit-frame-pointer -fno-optimize-sibling-calls -fsanitize=address -fsanitize=undefined -fsanitize-blacklist=/home/abdelhakeem/neovim/src/.asan-blacklist -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DEXITFREE -I/home/abdelhakeem/neovim/asan-build/config -I/home/abdelhakeem/neovim/src -I/home/abdelhakeem/neovim/.deps/usr/include -I/usr/include -I/home/abdelhakeem/neovim/asan-build/src/nvim/auto -I/home/abdelhakeem/neovim/asan-build/include
Compiled by abdelhakeem@abdelhakeem

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/share/nvim"

Run :checkhealth for more info
```
- Operating system/version: `Linux 5.1.15-arch1-1-ARCH #1 SMP PREEMPT Tue Jun 25 04:49:39 UTC 2019 x86_64 GNU/Linux`
- Terminal name/version: `# GNOME Terminal 3.32.2 using VTE 0.56.3 +GNUTLS`
- `$TERM`: `xterm-256color`

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
:au OptionSet * q
:silent! call nvim_create_buf(0, 1)
```

### Actual behaviour

```
=================================================================
==5078==ERROR: AddressSanitizer: heap-use-after-free on address 0x613000047c68 at pc 0x5584c43273b2 bp 0x7ffc6fad9c40 sp 0x7ffc6fad9c38
READ of size 4 at 0x613000047c68 thread T0
    #0 0x5584c43273b1 in hash_lock ~/neovim/asan-build/../src/nvim/hashtab.c:257:16
    #1 0x5584c3dcf1ea in vars_clear_ext ~/neovim/asan-build/../src/nvim/eval.c:20271:3
    #2 0x5584c3d68da9 in vars_clear ~/neovim/asan-build/../src/nvim/eval.c:20259:3
    #3 0x5584c506fe2a in win_free ~/neovim/asan-build/../src/nvim/window.c:4552:3
    #4 0x5584c503474f in win_free_mem ~/neovim/asan-build/../src/nvim/window.c:2678:3
    #5 0x5584c50369dc in win_free_all ~/neovim/asan-build/../src/nvim/window.c:2704:11
    #6 0x5584c45101c3 in free_all_mem ~/neovim/asan-build/../src/nvim/memory.c:647:3
    #7 0x5584c47e19a7 in mch_exit ~/neovim/asan-build/../src/nvim/os_unix.c:102:3
    #8 0x5584c43fc3a7 in getout ~/neovim/asan-build/../src/nvim/main.c:671:3
    #9 0x5584c413309b in ex_quit ~/neovim/asan-build/../src/nvim/ex_docmd.c:6107:7
    #10 0x5584c40c5585 in do_one_cmd ~/neovim/asan-build/../src/nvim/ex_docmd.c:2248:5
    #11 0x5584c40a8568 in do_cmdline ~/neovim/asan-build/../src/nvim/ex_docmd.c:594:20
    #12 0x5584c426b3a6 in apply_autocmds_group ~/neovim/asan-build/../src/nvim/fileio.c:6932:5
    #13 0x5584c4237355 in apply_autocmds ~/neovim/asan-build/../src/nvim/fileio.c:6562:10
    #14 0x5584c47786d5 in trigger_optionsset_string ~/neovim/asan-build/../src/nvim/option.c:4551:5
    #15 0x5584c4783d57 in set_string_option ~/neovim/asan-build/../src/nvim/option.c:2490:7
    #16 0x5584c473141b in set_option_value ~/neovim/asan-build/../src/nvim/option.c:4946:14
    #17 0x5584c3b5e82f in nvim_create_buf ~/neovim/asan-build/../src/nvim/api/vim.c:1000:5
    #18 0x5584c3a8fa67 in handle_nvim_create_buf ~/neovim/asan-build/src/nvim/auto/api/private/dispatch_wrappers.generated.h:3715:15
    #19 0x5584c3e41fa7 in api_wrapper ~/neovim/asan-build/../src/nvim/eval.c:6668:19
    #20 0x5584c3d8628c in call_func ~/neovim/asan-build/../src/nvim/eval.c:6536:11
    #21 0x5584c3d98676 in get_func_tv ~/neovim/asan-build/../src/nvim/eval.c:6277:11
    #22 0x5584c3d91831 in ex_call ~/neovim/asan-build/../src/nvim/eval.c:2887:9
    #23 0x5584c40c5585 in do_one_cmd ~/neovim/asan-build/../src/nvim/ex_docmd.c:2248:5
    #24 0x5584c40a8568 in do_cmdline ~/neovim/asan-build/../src/nvim/ex_docmd.c:594:20
    #25 0x5584c46496d1 in nv_colon ~/neovim/asan-build/../src/nvim/normal.c:4596:18
    #26 0x5584c4633f79 in normal_execute ~/neovim/asan-build/../src/nvim/normal.c:1131:3
    #27 0x5584c4cadc96 in state_enter ~/neovim/asan-build/../src/nvim/state.c:73:26
    #28 0x5584c45efbae in normal_enter ~/neovim/asan-build/../src/nvim/normal.c:463:3
    #29 0x5584c43e89ad in main ~/neovim/asan-build/../src/nvim/main.c:569:3
    #30 0x7f6ac0ce2ee2 in __libc_start_main (/usr/lib/libc.so.6+0x26ee2)
    #31 0x5584c38dfbdd in _start (~/neovim/asan-build/bin/nvim+0x9cfbdd)

0x613000047c68 is located 40 bytes inside of 352-byte region [0x613000047c40,0x613000047da0)
freed by thread T0 here:
    #0 0x5584c39c5b50 in __interceptor_free (~/neovim/asan-build/bin/nvim+0xab5b50)
    #1 0x5584c450d754 in xfree ~/neovim/asan-build/../src/nvim/memory.c:119:3
    #2 0x5584c3fa76e8 in tv_dict_free_dict ~/neovim/asan-build/../src/nvim/eval/typval.c:1374:3
    #3 0x5584c3fa77a6 in tv_dict_free ~/neovim/asan-build/../src/nvim/eval/typval.c:1387:5
    #4 0x5584c3fa7916 in tv_dict_unref ~/neovim/asan-build/../src/nvim/eval/typval.c:1400:5
    #5 0x5584c3dcf1c8 in unref_var_dict ~/neovim/asan-build/../src/nvim/eval.c:20249:3
    #6 0x5584c506fff8 in win_free ~/neovim/asan-build/../src/nvim/window.c:4554:3
    #7 0x5584c503474f in win_free_mem ~/neovim/asan-build/../src/nvim/window.c:2678:3
    #8 0x5584c5036963 in win_free_all ~/neovim/asan-build/../src/nvim/window.c:2700:11
    #9 0x5584c45101c3 in free_all_mem ~/neovim/asan-build/../src/nvim/memory.c:647:3
    #10 0x5584c47e19a7 in mch_exit ~/neovim/asan-build/../src/nvim/os_unix.c:102:3
    #11 0x5584c43fc3a7 in getout ~/neovim/asan-build/../src/nvim/main.c:671:3
    #12 0x5584c413309b in ex_quit ~/neovim/asan-build/../src/nvim/ex_docmd.c:6107:7
    #13 0x5584c40c5585 in do_one_cmd ~/neovim/asan-build/../src/nvim/ex_docmd.c:2248:5
    #14 0x5584c40a8568 in do_cmdline ~/neovim/asan-build/../src/nvim/ex_docmd.c:594:20
    #15 0x5584c426b3a6 in apply_autocmds_group ~/neovim/asan-build/../src/nvim/fileio.c:6932:5
    #16 0x5584c4237355 in apply_autocmds ~/neovim/asan-build/../src/nvim/fileio.c:6562:10
    #17 0x5584c47786d5 in trigger_optionsset_string ~/neovim/asan-build/../src/nvim/option.c:4551:5
    #18 0x5584c4783d57 in set_string_option ~/neovim/asan-build/../src/nvim/option.c:2490:7
    #19 0x5584c473141b in set_option_value ~/neovim/asan-build/../src/nvim/option.c:4946:14
    #20 0x5584c3b5e82f in nvim_create_buf ~/neovim/asan-build/../src/nvim/api/vim.c:1000:5
    #21 0x5584c3a8fa67 in handle_nvim_create_buf ~/neovim/asan-build/src/nvim/auto/api/private/dispatch_wrappers.generated.h:3715:15
    #22 0x5584c3e41fa7 in api_wrapper ~/neovim/asan-build/../src/nvim/eval.c:6668:19
    #23 0x5584c3d8628c in call_func ~/neovim/asan-build/../src/nvim/eval.c:6536:11
    #24 0x5584c3d98676 in get_func_tv ~/neovim/asan-build/../src/nvim/eval.c:6277:11
    #25 0x5584c3d91831 in ex_call ~/neovim/asan-build/../src/nvim/eval.c:2887:9
    #26 0x5584c40c5585 in do_one_cmd ~/neovim/asan-build/../src/nvim/ex_docmd.c:2248:5
    #27 0x5584c40a8568 in do_cmdline ~/neovim/asan-build/../src/nvim/ex_docmd.c:594:20
    #28 0x5584c46496d1 in nv_colon ~/neovim/asan-build/../src/nvim/normal.c:4596:18
    #29 0x5584c4633f79 in normal_execute ~/neovim/asan-build/../src/nvim/normal.c:1131:3

previously allocated by thread T0 here:
    #0 0x5584c39c5f4a in malloc (~/neovim/asan-build/bin/nvim+0xab5f4a)
    #1 0x5584c450d462 in try_malloc ~/neovim/asan-build/../src/nvim/memory.c:71:15
    #2 0x5584c450d61c in xmalloc ~/neovim/asan-build/../src/nvim/memory.c:105:15
    #3 0x5584c3fa4a1c in tv_dict_alloc ~/neovim/asan-build/../src/nvim/eval/typval.c:1307:21
    #4 0x5584c4ff9aae in win_alloc ~/neovim/asan-build/../src/nvim/window.c:4490:20
    #5 0x5584c4ff8e39 in win_new_float ~/neovim/asan-build/../src/nvim/window.c:562:10
    #6 0x5584c503c0da in win_alloc_aucmd_win ~/neovim/asan-build/../src/nvim/window.c:3494:15
    #7 0x5584c42468ff in aucmd_prepbuf ~/neovim/asan-build/../src/nvim/fileio.c:6402:5
    #8 0x5584c3b5e7f4 in nvim_create_buf ~/neovim/asan-build/../src/nvim/api/vim.c:999:5
    #9 0x5584c3a8fa67 in handle_nvim_create_buf ~/neovim/asan-build/src/nvim/auto/api/private/dispatch_wrappers.generated.h:3715:15
    #10 0x5584c3e41fa7 in api_wrapper ~/neovim/asan-build/../src/nvim/eval.c:6668:19
    #11 0x5584c3d8628c in call_func ~/neovim/asan-build/../src/nvim/eval.c:6536:11
    #12 0x5584c3d98676 in get_func_tv ~/neovim/asan-build/../src/nvim/eval.c:6277:11
    #13 0x5584c3d91831 in ex_call ~/neovim/asan-build/../src/nvim/eval.c:2887:9
    #14 0x5584c40c5585 in do_one_cmd ~/neovim/asan-build/../src/nvim/ex_docmd.c:2248:5
    #15 0x5584c40a8568 in do_cmdline ~/neovim/asan-build/../src/nvim/ex_docmd.c:594:20
    #16 0x5584c46496d1 in nv_colon ~/neovim/asan-build/../src/nvim/normal.c:4596:18
    #17 0x5584c4633f79 in normal_execute ~/neovim/asan-build/../src/nvim/normal.c:1131:3
    #18 0x5584c4cadc96 in state_enter ~/neovim/asan-build/../src/nvim/state.c:73:26
    #19 0x5584c45efbae in normal_enter ~/neovim/asan-build/../src/nvim/normal.c:463:3
    #20 0x5584c43e89ad in main ~/neovim/asan-build/../src/nvim/main.c:569:3
    #21 0x7f6ac0ce2ee2 in __libc_start_main (/usr/lib/libc.so.6+0x26ee2)

SUMMARY: AddressSanitizer: heap-use-after-free ~/neovim/asan-build/../src/nvim/hashtab.c:257:16 in hash_lock
Shadow bytes around the buggy address:
  0x0c2680000f30: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2680000f40: fd fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa
  0x0c2680000f50: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2680000f60: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2680000f70: 00 00 00 00 00 00 00 00 00 00 00 00 fa fa fa fa
=>0x0c2680000f80: fa fa fa fa fa fa fa fa fd fd fd fd fd[fd]fd fd
  0x0c2680000f90: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2680000fa0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2680000fb0: fd fd fd fd fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2680000fc0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2680000fd0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==5078==ABORTING
```

### Expected behaviour

Exit successfully without `heap-use-after-free`.

### PR DESCRIPTION
**Pull request by abdelhakeem**
Fixes #10838

### CODE REVIEW

File: src/nvim/window.c [hunk 1]
```diff
@@ -2694,17 +2694,18 @@ void win_free_all(void)
   while (first_tabpage->tp_next != NULL)
     tabpage_close(TRUE);
 
+  if (aucmd_win != NULL) {
+    win_remove(aucmd_win, NULL);
```
Comment (bfredl) [inline src/nvim/window.c]: I don't think `win_remove` is necessarily idempotent.  Could check if `wp->w_prev != NULL` before calling `win_remove (should be enough as aucmd_win cannot be first).

Comment (abdelhakeem) [reply src/nvim/window.c]: Doesn't `win_remove(aucmd_win, NULL)` guarantee that `aucmd_win` is not encountered in the `while (lastwin != NULL && lastwin->w_floating)` loop? It introduced more `heap-use-after-free` bugs though, the ASAN build fails on Travis in the `win_remove(aucmd_win, NULL)` call when `aucmd_win` is manually closed before. Am I missing something?

Comment (bfredl) [reply src/nvim/window.c]: No, I just mean the aucmd_win might already (actually typically) be removed before _this_ line. You have to check if it is in the list before calling `win_remove`.

Comment (abdelhakeem) [reply src/nvim/window.c]: Oh, yes, that's why it fails in some tests

**Pull request merged**